### ISSUE TITLE
Styled HoC

### ISSUE THREAD

**Issue by kachkaev** [styled-components/styled-components#316]
I've probably just missed this in the docs, but still decided to create this issue to help others find it.

I'm wondering if it is possible to create a styled HoC like this:  
(the example does not work, it just shows the idea)

```js
import styled from 'styled-componets';
import { compose, pure } from 'recompose';
import withXYZ from '../hocs/withXYZ';

function MyBareboneComponent({ className, x, y, z }) {
  return (
    <div className={className}>
      {x} ... {y} ... {z}
    </div>
  )
}

const withStyles = styled`
    color: ${x === 'abc' ? 'red' : 'black'};
`;

export default compose(
  withXYZ,
  pure,
  withStyles,
)(MyBareboneComponent);
```

HoC `withStyles` will just inject styles to whatever component it's being used with. In a perfect world this HoC would be compatible with [`recompose`](https://github.com/acdlite/recompose)'s `compose` and shrink into one wrapper component together with other HoCs that are added to `MyBareboneComponent`.

What do you think?

**Comment by k15a** [comment]
You can do: ```styled(MyBareboneComponent)```
I think a styled HOC wouldn't be a good addition because styled components is all about removing the mapping between styles and the components. A HOC would mean that you need to map the styles to the component yourself.

**Comment by kachkaev** [comment]
I understand that the main idea of styled components is keeping them close to the components they wrap. HoC option just let's you do what you've already been doing, but using a slightly different notation. Now I'd have to write things like this - same thing, but harder to read:

```js
import styled from 'styled-componets';
import { compose, pure } from 'recompose';
import withXYZ from './withXYZ';

function MyBareboneComponent({ className, x, y, z }) {
  return (
    <div className={className}>
      {x} ... {y} ... {z}
    </div>
  )
}

const MyBareboneComponentwithStyles = styled(MyBareboneComponent)`
    color: ${x === 'abc' ? 'red' : 'black'};
`;

export default compose(
  withXYZ,
  pure,
)(MyBareboneComponentwithStyles);
```

HoC option allows you to place styles on top of the component definition if you want + if it's properly implemented, it can be combined with other HoCs and then flattened into a single wrapper for `MyBareboneComponentwithStyles` by `compose`. This will make React tree less bulky and thus can have a positive performance impact.

**Comment by kachkaev** [comment]
Check out this talk by the author of `recompose`: https://www.youtube.com/watch?v=zD_judE-bXk

**Comment by diegohaz** [comment]
I'm not very familiar with `recompose`, but does this work for you?
```jsx
import React from 'react';
import styled from 'styled-components';

const withX = (WrappedComponent) => (props) =>
  <WrappedComponent {...props} x="abc" />

const withStyles = (WrappedComponent) => styled(WrappedComponent)`
  color: ${props => props.x === 'abc' ? 'red' : 'black'};
`

const HelloWorld = ({ className }) => 
  <div className={className}>HelloWorld</div>

export default withX(withStyles(HelloWorld));
```

**Comment by kachkaev** [comment]
Looking at the code you shared @diegohaz, I think it'd do the same. But it's still just a longer notation of the original example:

```js
const withStyles = (WrappedComponent) => styled(WrappedComponent)`
 ...
`;
// vs.
const withStyles = styled`
 ...
`;
```
:–)

Actually it'd be probably complicated and unnecessary to make `styled` a function that takes both a component and an array of arguments and acts differently based on that. So the HoC generator could be named, let's say `styledHOC`.

What other technical issues might be there? Maybe the whole idea of a styled HOC is not in line with the library internals, I just don't see this from a distance. Or maybe I'm missing something else.

**Comment by diegohaz** [comment]
`styled` itself is a sort of HOC. You can do something like this (actually, this code works), but it seems a little redundant for me:
```jsx
import React from 'react';
import styled from 'styled-components';

const styledHOC = (strings, ...interpolations) => (WrappedComponent) => 
  styled(WrappedComponent)(strings, ...interpolations)

const withX = (WrappedComponent) => (props) =>
  <WrappedComponent {...props} x="abc" />

const withStyles = styledHOC`
  color: ${props => props.x === 'abc' ? 'red' : 'black'};
`

const HelloWorld = ({ className }) => 
  <div className={className}>HelloWorld</div>

export default withX(withStyles(HelloWorld));
```

**Comment by kachkaev** [comment]
The last line seems quite like what I was after. But is not this just a reverse action of what's happening in [`src/constructors/styled.js`](https://github.com/styled-components/styled-components/blob/f7b608113e663b5c39fd48701dba8913fe381603/src/constructors/styled.js)? 

Could not just [`src/models/StyledComponent.js`](https://github.com/styled-components/styled-components/blob/f7b608113e663b5c39fd48701dba8913fe381603/src/models/StyledComponent.js) or [`src/models/StyledNativeComponent.js`](https://github.com/styled-components/styled-components/blob/f7b608113e663b5c39fd48701dba8913fe381603/src/models/StyledNativeComponent.js) be exported by the module as `styledHOC` based on the current environment?

**Comment by mxstbr** [comment]
I don't really see the value in this at all. Going back to the original example, this is what you did:

```JSX

    <div className={className}>
      {x} ... {y} ... {z}
    </div>
```

This `className` is then injected via `recompose` and HoC from the styles further below:

```JS
const withStyles = styled`
    color: ${x === 'abc' ? 'red' : 'black'};
`;
```

Why do you not just create a styled component for that? What's the benefit of having a HoC?

```JS
const Wrapper = styled.div`
  color: ${x === 'abc' ? 'red' : 'black'};
`
function MyBareboneComponent({ x, y, z }) {
  return (
    <Wrapper>
      {x} ... {y} ... {z}
    </Wrapper>
  )
}
```

Exact same functionality, works perfectly fine today and gets rid of a bunch of complexity because there's no need for a specific HoC. What am I missing here?

**Comment by kachkaev** [comment]
Hey Max! Here are a couple of possible points:

1. You might want to have something like this:

```js
const withStyles = styledHOC`
   color: red;
   &__x {
      padding: 42px;
   }
   &__y {
      color: ${y === 'error' ? 'red: 'green'};
   }
`
```

2. You might want to make

**Comment by mxstbr** [comment]
> You might want to have something like this:

The way you've done this is definitely not something we want to encourage. On top of that, this can easily be achieved like so:

```JSX
const Wrapper = styled.div`
  color: red;
`

const X = styled.span`
  padding: 42px;
`

const Y = styled.span`
  color: ${props => props.error ? 'red': 'green'};
`

const Z = styled.span`
  display: none;
`

function MyBareboneComponent({ x, y, z, className }) {
  return (
    <Wrapper>
      <X>{x}</A>
      <Y error={y === 'error'}>{y}</Y>
      <Z>{z}</Z>
    </div>
  );
}
```

That's idiomatic usage of `styled-components`!

I'm not sure I fully understand the second point, what are those recompose functions doing that styled-components and/or a `shouldComponentUpdate` don't?

**Comment by kachkaev** [comment]
Those HOCs in `recompose` do nothing special really, they just encourage a clear distinction between representation components and containers of any kind. Rather than switching between a function and a class just to make the component pure, you `export  pure(MyFunctionalComponent)`. Any custom behaviour of that functional component depends on HOCs into which the component is wrapped: `withData(withSomeStuffFromContext(pure(MyFunctionalComponent)))`, which is the same as `compose(withData, withSomeStuffFromContext, pure)(MyFunctionalComponent)))`. When using `compose`, it is in theory possible to flatten those three HOCs into a single one and thus make the resulting tree not so bulky. [Andrew Clark's talks at this year's React Europe](https://www.youtube.com/watch?v=zD_judE-bXk) explains the rational of all this.

I personally find defining these extra `X`, `Y` anx `Z` too verbose in the situation above. What will happen to my app's performance if I have a 100 instances of `MyBareboneComponent`?

**Comment by k15a** [comment]
You can actually write this HOC in one line (didn't test it). So it shouldn't be too hard to copy - paste this to every file where you need the HOC or make a small module like `styled-components-hoc`.

```JS
const ws = (...args) => (Component) => styled(Component)(...args)

// Usage
const withStyle = ws`
	/* Your CSS */
`

export default compose(
   withXYZ,
   onlyUpdateForKeys(['x', 'y', 'z']),
   withStyles,
)(MyBareboneComponent)
```

IMHO the extra `X`, `Y` and `Z` are way easier to read than  ```<span className={`${className}__x`}>{x}</span>```

**Comment by kachkaev** [comment]
@k15a yeah I know I can use `const ws = (...args) => (Component) => styled(Component)(...args)` as @diegohaz has suggested this above. This is what I'm likely to use now to maintain coherence with other HOCs. The reason is that treating `className` as just another prop from just another HOC is cognitively easier to me than treating it as some special case.

What is funny is that this `ws` function would be a HOC that turns a non-HOC, which is by itself a wrapper for a HOC, back into a HOC :-D The `styled-components` module does have this `ws` function internally in some form already; that is [`src/models/StyledComponent.js`](https://github.com/styled-components/styled-components/blob/f7b608113e663b5c39fd48701dba8913fe381603/src/models/StyledComponent.js) and [`src/models/StyledNativeComponent.js`](https://github.com/styled-components/styled-components/blob/f7b608113e663b5c39fd48701dba8913fe381603/src/models/StyledNativeComponent.js).

My overall point is that it'd be great if this `ws` function was exported by `styled-components` rather than coming from a wrapper module `styled-components-hoc` (because a wrapper module will be unwrapping a wrapper) :–)

**Comment by k15a** [comment]
Ahh sorry I overlooked that comment.

**Comment by mxstbr** [comment]
> What will happen to the app's performance if I have a 100 instances of this button that's actually by itself a pretty large component tree (because we created X, Y and Z)?

Absolutely nothing.

If your app has performance problems then I'm pretty certain it's not because of too many components. Components literally are React. (imo, if too many components are a performance problem in your app, React isn't a good fit)

> I personally find defining these extra X, Y anx Z too verbose in the situation above. It looks harder to read to me plus there is too much to refactor if something needs to be changed inside MyBareboneComponent.

That's just personal preference and not a very good reason to export a HoC that encourages to breaks the whole premise of removing the mapping between styles and components.

What I recommend doing (and what I do) is have those tiny components in extra files:

```JS
// MyComp/Wrapper.js
export default styled.div`
  color: red;
`
```

```JS
// MyComp/X.js
export default styled.span`
  padding: 42px;
`
```


```JS
// MyComp/Y.js
export default styled.span`
  color: ${props => props.error ? 'red': 'green'};
`
```

```JS
// MyComp/Z.js
export default styled.span`
  display: none;
`
```

```JSX
// MyComp/index.js
import X from './X.js'
import Y from './Y.js'
import Z from './Z.js'

function MyBareboneComponent({ x, y, z, className }) {
  return (
    <Wrapper>
      <X>{x}</X>
      <Y error={y === 'error'}>{y}</Y>
      <Z>{z}</Z>
    </Wrapper>
  );
}
```

*See the [postcss.parts](https://github.com/mxstbr/postcss.parts) codebase for an example of a real codebase that uses s-c and this convention*

That way the main component file stays small and is, in fact, easier to read than before! Contrary to the original `<span className={`${className}__x`} />` soup, it now just says `<X>`. (substitute with semantic names, of course, not `"X"`)

> React component != DOM node; it is something semantically meaningful, reusable and testable

The core idea behind styled-components is that you couple DOM nodes to their styles to enforce the presentational/container component split. That styled components are React components under the hood is an _implementation detail_, not something you are concerned with. Treat them like any other DOM node!

**Comment by kachkaev** [comment]
Thanks a lot for sharing your thoughts guys and for arguing with me about this suggestion! I've recently switched to using something like

```jsx
    <Wrapper>
      <X>{x}</X>
      <Y error={y === 'error'}>{y}</Y>
      <Z>{z}</Z>
    </Wrapper>
```

and I'm really enjoying it! Indeed, there seems to be no need for `withStyles` HOC!

I still believe that there might appear some performance issues if the component tree grows too big. But this has chances to get solved with [react-fiber](http://isfiberreadyyet.com/) before I even notice any problems :–)

**Comment by mxstbr** [comment]
> I still believe that there might appear some performance issues if the component tree grows too big. 

We've had lots of people be concerned about this, but nobody ever reporting actual performance issues in production :wink: obviously YMMV, but even if you run into troubles there we'll do our very best to work on that part.

**Comment by dmk255** [comment]
Here's a potential use case for a styled HOC though it may be slightly different than above.

My team is working on developing a tool that will be embedded on our clients websites. These websites can have terrible css that will conflict with our embed. I've been able to get around this by making my styled components more specific:

```JS
const MyComponent = styled.div`
&&& {
height: 100px;
width: 100px;
background: black
}
`
```

Results in a `.Zsc12c.Zsc12c.Zsc12c` selector bumping up specificity. 

The issue I'm running into is that i's very annoying to have to remember to place `&&& { ... }` on __every__ styled component (some people have a generic apply css to every `div` which breaks our stuff since tags have higher specificity than classnames).

I've been able to get around this in most cases with the following HOC.
```JS
const styledHOC = Component => (strings, ...args) => {
  const wrappedCSS = ['&& {', ...strings, '}'].join('\n');
  return styled(Component)([wrappedCSS], ...args);
};
```

However, this HOC only works in the simple case where we're just applying styled to a component. We have other cases where we call `styledHOC(MyComponent).attrs({...})'...'` and this HOC breaks since we don't expose the styled API on the first call.

There may be an easy solution I'm missing but I would love to have a way to wrap all of my styled components to increase their specificity while preserving the styled API.
